src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs

```diff
@@ -324,17 +324,23 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
     /// <param name="context">The <see cref="JsonSchemaExporterContext"/> associated with the current type.</param>
     internal static void ApplyPolymorphismOptions(this JsonNode schema, JsonSchemaExporterContext context)
     {
-        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions)
+        // The `context.Path.Length == 0` check is used to ensure that we only apply the polymorphism options
+        // to the top-level schema and not to any nested schemas that are generated.
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.Path.Length == 0)
         {
             var mappings = new JsonObject();
             foreach (var derivedType in polymorphismOptions.DerivedTypes)
             {
-                if (derivedType.TypeDiscriminator is null)
+                if (derivedType.TypeDiscriminator is { } discriminator)
                 {
-                    continue;
+                    var jsonDerivedType = context.TypeInfo.Options.GetTypeInfo(derivedType.DerivedType);
+                    // Discriminator mappings are only supported in OpenAPI v3+ so we can safely assume that
```

Comment:
We don't version check ourselves. The in-memory model doesn't have the concept of a version. Instead, OpenAPI.NET will determine whether or not this property is going to be serialized when it writes the document out (see [here](https://github.com/microsoft/OpenAPI.NET/blob/4fe58269277ee939775ec98e0ffa7b5706c958bc/src/Microsoft.OpenApi/Models/OpenApiDiscriminator.cs#L42-L63)). We could version check ourselves here but it would require pulling the info from `OpenApiOptions` and I'd rather take advantage of what OpenApi.NET does here.